# Studi Kasus Data Cleansing dan Data Enrichment
## Dataset Lagu (Musik)

**Mata Kuliah:** Data Mining, Pertemuan ke-4 (Belajar Mandiri Asinkron)
**Nama:** David Andrew Gregorius Tingang Liah
**NIM:** 2418058
**Kelas:** C
**Program Studi:** Teknik Informatika, ITN Malang

---

Di tugas ini saya memakai dataset kecil tentang lagu-lagu populer (35 baris). Datasetnya saya susun sendiri dalam bentuk file CSV, jadi nilai-nilainya (jumlah putar, popularitas, durasi, dan tanggal) adalah data contoh untuk latihan dan bukan angka resmi dari platform musik mana pun. Datanya memang sengaja saya buat "kotor" supaya berbagai teknik cleansing bisa dipraktikkan, dan masalahnya mirip dengan yang sering muncul di data nyata.

Alur pengerjaan:
1. Memahami dataset
2. Mengidentifikasi masalah kualitas data
3. Data Cleansing
4. Data Enrichment
5. Perbandingan kondisi data sebelum dan sesudah
6. Kesimpulan

## 0. Import Library

In [1]:
import io
from datetime import datetime

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 200)

## 1. Pemahaman Dataset

Data mentah saya tulis langsung di dalam notebook (sebagai teks CSV) supaya notebook ini bisa dijalankan di Colab tanpa perlu upload file. File yang sama juga ada di GitHub dengan nama `dataset_musik_mentah.csv`.

In [2]:
data_mentah = """id_lagu,judul,artis,genre,tahun_rilis,tanggal_rilis,durasi,jumlah_putar,popularitas
1,Shape of You,Ed Sheeran,Pop,2017.0,2017-01-06,3:53,"3,100,000,000",90.0
2,Blinding Lights,THE WEEKND,pop,2019.0,29/11/2019,3:20,"3,400,000,000",92.0
3,Bohemian Rhapsody,Queen,Rock,1975.0,1975-10-31,5:55,"2,000,000,000",88.0
4,Someone Like You,Adele,Pop,2011.0,24/01/2011,4:45,"1,700,000,000",80.0
5,Lose Yourself,Eminem,hip hop,2002.0,2002-10-28,5:26,"2,100,000,000",85.0
6,Hotel California,Eagles,Rock ,1976.0,08-12-1976,6:30,"1,500,000,000",82.0
7,Smells Like Teen Spirit,NIRVANA,Rock,1991.0,1991-09-10,5:01,"1,800,000,000",83.0
8,Billie Jean,Michael Jackson,POP,1983.0,02/01/1983,4:54,"1,600,000,000",84.0
9,Uptown Funk,Mark Ronson,Funk,2014.0,2014-11-10,4:30,"2,300,000,000",81.0
10,Despacito,Luis Fonsi,Latin,2017.0,12/01/2017,228,"3,000,000,000",78.0
11,Believer,Imagine Dragons,Rock,2017.0,2017-02-01,3:24,"2,300,000,000",
12,Bad Guy,Billie Eilish,Pop,2019.0,2019-03-29,3:14,"2,800,000,000",87.0
13,Dynamite,BTS,Kpop,2020.0,21/08/2020,3:19,"1,400,000,000",82.0
14,Lathi,Weird Genius,EDM,2020.0,2020-02-28,3:33,"650,000,000",75.0
15,Hati-Hati di Jalan,Tulus,Pop,2022.0,2022-02-18,250,"520,000,000",74.0
16,Kangen,Dewa 19,Rock,1992.0,1992-03-15,58:20,"410,000,000",70.0
17,Yellow,Coldplay,,2000.0,2000-06-26,4:29,"1,900,000,000",84.0
18,Shake It Off,Taylor Swift,Pop,2014.0,18/08/2014,3:39,"2,000,000,000",86.0
19,Just the Way You Are,Bruno Mars,Pop,2010.0,2010-07-20,3:40,"1,800,000,000",-5.0
20,Levitating,Dua Lipa,Pop,2035.0,2020-03-27,3:23,"2,500,000,000",88.0
21,As It Was,Harry Styles,pop,2022.0,31/03/2022,2:37,"3,300,000,000",91.0
22,Umbrella,Rihanna,Pop,2007.0,2007-03-29,4:36,"1,500,000,000",150.0
23,Sunday Morning,Maroon 5,Pop,2004.0,,4:01,"1,100,000,000",76.0
24,In the End,Linkin Park,Rock,2001.0,09/10/2001,3:36,"1,900,000,000",83.0
25,Creep,Radiohead,Rock,1992.0,21/09/1992,3:58,"1,400,000,000",78.0
26,Sephia,Sheila on 7,Pop,1999.0,31/02/1999,243,"380,000,000",68.0
27,Dancing Queen,ABBA,Disco,1976.0,1976-08-16,3:51,"1,500,000,000",80.0
28,Wake Me Up,Avicii,EDM,,2013-06-17,4:07,"1,800,000,000",82.0
29,Get Lucky,Daft Punk,Disco,2013.0,2013-04-19,6:09,"-1,000,000,000",79.0
30,Gangnam Style,PSY,K-Pop,2012.0,2012-07-15,3:39,"95,000,000,000",77.0
31,Kill This Love,Blackpink,K-pop,2019.0,2019-04-04,3:11,"1,000,000,000",81.0
32,Someone You Loved,Lewis Capaldi,Pop,2018.0,2018-11-08,3:02,,85.0
33,Separuh Aku,Noah,Pop,2012.0,2012-05-20,0:00,"300,000,000",
4,Someone Like You,Adele,Pop,2011.0,24/01/2011,4:45,"1,700,000,000",80.0
10, DESPACITO ,luis fonsi,latin,2017.0,12/01/2017,228,"3,000,000,000",78.0"""

df_mentah = pd.read_csv(io.StringIO(data_mentah))
df_mentah.head(10)

,id_lagu,judul,artis,genre,tahun_rilis,tanggal_rilis,durasi,jumlah_putar,popularitas
0,1,Shape of You,Ed Sheeran,Pop,2017.0,2017-01-06,3:53,"3,100,000,000",90.0
1,2,Blinding Lights,THE WEEKND,pop,2019.0,29/11/2019,3:20,"3,400,000,000",92.0
2,3,Bohemian Rhapsody,Queen,Rock,1975.0,1975-10-31,5:55,"2,000,000,000",88.0
3,4,Someone Like You,Adele,Pop,2011.0,24/01/2011,4:45,"1,700,000,000",80.0
4,5,Lose Yourself,Eminem,hip hop,2002.0,2002-10-28,5:26,"2,100,000,000",85.0
5,6,Hotel California,Eagles,Rock,1976.0,08-12-1976,6:30,"1,500,000,000",82.0
6,7,Smells Like Teen Spirit,NIRVANA,Rock,1991.0,1991-09-10,5:01,"1,800,000,000",83.0
7,8,Billie Jean,Michael Jackson,POP,1983.0,02/01/1983,4:54,"1,600,000,000",84.0
8,9,Uptown Funk,Mark Ronson,Funk,2014.0,2014-11-10,4:30,"2,300,000,000",81.0
9,10,Despacito,Luis Fonsi,Latin,2017.0,12/01/2017,228,"3,000,000,000",78.0


In [3]:
print("Jumlah baris dan kolom:", df_mentah.shape)
print()
df_mentah.info()

Jumlah baris dan kolom: (35, 9)

<class 'pandas.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_lagu        35 non-null     int64  
 1   judul          35 non-null     str    
 2   artis          35 non-null     str    
 3   genre          34 non-null     str    
 4   tahun_rilis    34 non-null     float64
 5   tanggal_rilis  34 non-null     str    
 6   durasi         35 non-null     str    
 7   jumlah_putar   34 non-null     str    
 8   popularitas    33 non-null     float64
dtypes: float64(2), int64(1), str(6)
memory usage: 2.6 KB


Penjelasan kolom pada dataset:

| Kolom | Keterangan |
|---|---|
| id_lagu | nomor identitas lagu |
| judul | judul lagu |
| artis | nama artis atau band |
| genre | genre lagu |
| tahun_rilis | tahun lagu dirilis |
| tanggal_rilis | tanggal lagu dirilis |
| durasi | lama lagu (menit:detik, tapi ada yang ditulis dalam detik) |
| jumlah_putar | jumlah lagu diputar (streaming) |
| popularitas | skor popularitas, seharusnya antara 0 sampai 100 |

Dari `info()` kelihatan beberapa kolom tipenya belum sesuai. `jumlah_putar`, `durasi`, dan `tanggal_rilis` terbaca sebagai teks (object), padahal seharusnya angka dan tanggal. Kolom `tahun_rilis` dan `popularitas` terbaca sebagai desimal (float), itu tanda ada nilai yang kosong.

## 2. Identifikasi Masalah Kualitas Data

Di tahap ini saya hanya memeriksa dan mencatat masalahnya dulu, belum ada data yang diubah.

### 2.1 Missing value

In [4]:
missing = df_mentah.isna().sum()
print("Jumlah nilai kosong per kolom:")
print(missing[missing > 0])
print()
print("Total nilai kosong:", missing.sum())

Jumlah nilai kosong per kolom:
genre            1
tahun_rilis      1
tanggal_rilis    1
jumlah_putar     1
popularitas      2
dtype: int64

Total nilai kosong: 6


In [5]:
# baris yang punya nilai kosong
df_mentah[df_mentah.isna().any(axis=1)]

,id_lagu,judul,artis,genre,tahun_rilis,tanggal_rilis,durasi,jumlah_putar,popularitas
10,11,Believer,Imagine Dragons,Rock,2017.0,2017-02-01,3:24,"2,300,000,000",NaN
16,17,Yellow,Coldplay,NaN,2000.0,2000-06-26,4:29,"1,900,000,000",84.0
22,23,Sunday Morning,Maroon 5,Pop,2004.0,NaN,4:01,"1,100,000,000",76.0
27,28,Wake Me Up,Avicii,EDM,NaN,2013-06-17,4:07,"1,800,000,000",82.0
31,32,Someone You Loved,Lewis Capaldi,Pop,2018.0,2018-11-08,3:02,NaN,85.0
32,33,Separuh Aku,Noah,Pop,2012.0,2012-05-20,0:00,"300,000,000",NaN


### 2.2 Data duplikat

In [6]:
print("Duplikat persis (semua kolom sama):", df_mentah.duplicated().sum())

# duplikat yang "tersembunyi": isinya sama tapi huruf besar/kecil atau spasinya beda
kunci = df_mentah["judul"].str.strip().str.lower() + "|" + df_mentah["artis"].str.strip().str.lower()
print("Duplikat setelah huruf dan spasi disamakan:", kunci.duplicated().sum())
print()
df_mentah[kunci.duplicated(keep=False)]

Duplikat persis (semua kolom sama): 1
Duplikat setelah huruf dan spasi disamakan: 2



,id_lagu,judul,artis,genre,tahun_rilis,tanggal_rilis,durasi,jumlah_putar,popularitas
3,4,Someone Like You,Adele,Pop,2011.0,24/01/2011,4:45,"1,700,000,000",80.0
9,10,Despacito,Luis Fonsi,Latin,2017.0,12/01/2017,228,"3,000,000,000",78.0
33,4,Someone Like You,Adele,Pop,2011.0,24/01/2011,4:45,"1,700,000,000",80.0
34,10,DESPACITO,luis fonsi,latin,2017.0,12/01/2017,228,"3,000,000,000",78.0


Ada 1 duplikat yang persis sama (Someone Like You) dan 1 lagi yang baru ketahuan kalau huruf besar-kecil dan spasinya disamakan (Despacito). Jadi total ada 2 baris duplikat.

### 2.3 Inkonsistensi format

In [7]:
print("Nilai unik kolom genre:")
print(df_mentah["genre"].unique())

Nilai unik kolom genre:
<StringArray>
['Pop', 'pop', 'Rock', 'hip hop', 'Rock ', 'POP', 'Funk', 'Latin', 'Kpop', 'EDM', nan, 'Disco', 'K-Pop', 'K-pop', 'latin']
Length: 15, dtype: str


In [8]:
print("Nilai unik kolom durasi:")
print(df_mentah["durasi"].unique())

Nilai unik kolom durasi:
<StringArray>
[ '3:53',  '3:20',  '5:55',  '4:45',  '5:26',  '6:30',  '5:01',  '4:54',  '4:30',   '228',  '3:24',  '3:14',  '3:19',  '3:33',   '250', '58:20',  '4:29',  '3:39',  '3:40',  '3:23',  '2:37',  '4:36',
  '4:01',  '3:36',  '3:58',   '243',  '3:51',  '4:07',  '6:09',  '3:11',  '3:02',  '0:00']
Length: 32, dtype: str


In [9]:
print("Nilai unik kolom tanggal_rilis:")
print(df_mentah["tanggal_rilis"].dropna().unique())

Nilai unik kolom tanggal_rilis:
<StringArray>
['2017-01-06', '29/11/2019', '1975-10-31', '24/01/2011', '2002-10-28', '08-12-1976', '1991-09-10', '02/01/1983', '2014-11-10', '12/01/2017', '2017-02-01', '2019-03-29', '21/08/2020', '2020-02-28',
 '2022-02-18', '1992-03-15', '2000-06-26', '18/08/2014', '2010-07-20', '2020-03-27', '31/03/2022', '2007-03-29', '09/10/2001', '21/09/1992', '31/02/1999', '1976-08-16', '2013-06-17', '2013-04-19',
 '2012-07-15', '2019-04-04', '2018-11-08', '2012-05-20']
Length: 32, dtype: str


In [10]:
# cek spasi berlebih dan huruf kapital yang tidak konsisten
for kol in ["judul", "artis", "genre"]:
    teks = df_mentah[kol].dropna()
    ada_spasi = (teks != teks.str.strip()).sum()
    semua_besar = teks.apply(lambda x: x.strip().isupper()).sum()
    semua_kecil = teks.apply(lambda x: x.strip().islower()).sum()
    print(f"{kol}: spasi berlebih={ada_spasi}, huruf besar semua={semua_besar}, huruf kecil semua={semua_kecil}")

judul: spasi berlebih=1, huruf besar semua=1, huruf kecil semua=0
artis: spasi berlebih=0, huruf besar semua=5, huruf kecil semua=1
genre: spasi berlebih=1, huruf besar semua=3, huruf kecil semua=4


Genre sama yang ditulis beda-beda: `Pop`, `pop`, `POP`, `Rock` dan `Rock ` (ada spasi di belakang), `hip hop`, `Kpop`, `K-Pop`, `K-pop`. Format durasi juga campur, ada yang `3:53` (menit:detik) dan ada yang langsung dalam detik seperti `228`. Format tanggal ada tiga macam (`2017-01-06`, `24/01/2011`, `08-12-1976`).

### 2.4 Kesalahan tipe data

Kolom `jumlah_putar` berisi angka yang memakai tanda koma ribuan, jadi terbaca sebagai teks. Kolom `durasi` dan `tanggal_rilis` juga teks. Ini perlu diubah supaya bisa dihitung.

### 2.5 Data yang tidak valid

In [11]:
# fungsi bantu: ubah durasi (m:dd atau detik) menjadi detik
def durasi_ke_detik(x):
    if pd.isna(x):
        return np.nan
    x = str(x).strip()
    if ":" in x:
        menit, detik = x.split(":")
        return int(menit) * 60 + int(detik)
    return float(x)


putar_cek = pd.to_numeric(df_mentah["jumlah_putar"].str.replace(",", ""), errors="coerce")
detik_cek = df_mentah["durasi"].apply(durasi_ke_detik)

print("Popularitas di luar rentang 0-100:")
print(df_mentah[(df_mentah["popularitas"] < 0) | (df_mentah["popularitas"] > 100)][["judul", "popularitas"]])
print()
print("Tahun rilis tidak masuk akal (di luar 1950-2026):")
print(df_mentah[(df_mentah["tahun_rilis"] < 1950) | (df_mentah["tahun_rilis"] > 2026)][["judul", "tahun_rilis"]])
print()
print("Jumlah putar negatif:")
print(df_mentah[putar_cek < 0][["judul", "jumlah_putar"]])
print()
print("Durasi 0 detik:")
print(df_mentah[detik_cek < 30][["judul", "durasi"]])
print()
print("Tanggal yang tidak ada di kalender:")
print(df_mentah[df_mentah["tanggal_rilis"] == "31/02/1999"][["judul", "tanggal_rilis"]])

Popularitas di luar rentang 0-100:
                   judul  popularitas
18  Just the Way You Are         -5.0
21              Umbrella        150.0

Tahun rilis tidak masuk akal (di luar 1950-2026):
         judul  tahun_rilis
19  Levitating       2035.0

Jumlah putar negatif:
        judul    jumlah_putar
28  Get Lucky  -1,000,000,000

Durasi 0 detik:
          judul durasi
32  Separuh Aku   0:00

Tanggal yang tidak ada di kalender:
     judul tanggal_rilis
25  Sephia    31/02/1999


### 2.6 Outlier

Untuk outlier saya pakai metode IQR (batas bawah = Q1 - 1.5 x IQR, batas atas = Q3 + 1.5 x IQR). Nilai yang ditandai IQR tidak langsung dibuang, saya lihat dulu satu per satu apakah memang salah atau cuma nilai yang besar tapi wajar.

In [12]:
def batas_iqr(seri):
    q1 = seri.quantile(0.25)
    q3 = seri.quantile(0.75)
    iqr = q3 - q1
    return q1 - 1.5 * iqr, q3 + 1.5 * iqr


# hanya nilai yang masuk akal dulu (bukan negatif / 0) supaya batasnya tidak kacau
detik_valid = detik_cek[detik_cek >= 30]
bawah, atas = batas_iqr(detik_valid)
print(f"Durasi (detik): batas bawah={bawah:.0f}, batas atas={atas:.0f}")
print(df_mentah.loc[(detik_cek > atas) | ((detik_cek < bawah) & (detik_cek >= 30)), ["judul", "durasi"]])
print()

putar_valid = putar_cek[putar_cek > 0]
bawah, atas = batas_iqr(putar_valid)
print(f"Jumlah putar: batas bawah={bawah:,.0f}, batas atas={atas:,.0f}")
print(df_mentah.loc[(putar_cek > atas) | ((putar_cek < bawah) & (putar_cek > 0)), ["judul", "jumlah_putar"]])

Durasi (detik): batas bawah=110, batas atas=386
               judul durasi
5   Hotel California   6:30
15            Kangen  58:20

Jumlah putar: batas bawah=50,000,000, batas atas=3,650,000,000
            judul    jumlah_putar
29  Gangnam Style  95,000,000,000


Hasil pengecekan:
- **Durasi:** IQR menandai dua lagu, yaitu `58:20` (Kangen) dan `6:30` (Hotel California). Yang 58 menit jelas tidak wajar untuk satu lagu, jadi itu saya anggap salah. Hotel California memang lagu yang panjang tapi 6 menit lebih masih wajar, jadi tidak saya ubah.
- **Jumlah putar:** `95.000.000.000` (Gangnam Style) jauh di atas yang lain. Itu sepertinya salah ketik (kelebihan nol), jadi saya anggap outlier yang salah. Lagu yang jumlah putarnya besar tapi wajar tidak saya ubah.

### Ringkasan masalah yang ditemukan

| No | Jenis masalah | Kolom | Jumlah |
|---|---|---|---|
| 1 | Missing value | tahun_rilis, tanggal_rilis, genre, jumlah_putar, popularitas | 6 sel kosong |
| 2 | Data duplikat | seluruh baris | 2 baris |
| 3 | Inkonsistensi format | genre (huruf besar/kecil, spasi, penulisan) | 14 penulisan untuk 8 genre |
| 4 | Inkonsistensi format | durasi (menit:detik dan detik) | 4 baris dalam detik |
| 5 | Inkonsistensi format | tanggal_rilis (3 format) | 3 format |
| 6 | Inkonsistensi teks | judul dan artis (huruf besar semua / kecil semua, spasi) | 3 baris |
| 7 | Kesalahan tipe data | jumlah_putar, durasi, tanggal_rilis (terbaca teks) | 3 kolom |
| 8 | Data tidak valid | popularitas (-5 dan 150), tahun_rilis (2035), jumlah_putar (negatif), durasi (0:00), tanggal (31/02) | 6 nilai |
| 9 | Outlier | durasi (58:20), jumlah_putar (95 miliar) | 2 nilai |

## 3. Data Cleansing

Pertama saya salin dulu data mentahnya ke variabel `df`, supaya data asli (`df_mentah`) tetap utuh dan bisa dibandingkan nanti.

Urutan pembersihan saya atur begini: teks dirapikan dulu, lalu tipe data diperbaiki, lalu nilai tidak valid dijadikan kosong, lalu duplikat dibuang, lalu outlier ditangani, dan terakhir baru nilai kosong diisi. Duplikat dibuang sebelum pengisian supaya nilai median tidak terpengaruh baris ganda.

In [13]:
df = df_mentah.copy()
print("Jumlah baris awal:", len(df))

Jumlah baris awal: 35


### 3.1 Standardisasi teks (judul, artis, genre)

In [14]:
# buang spasi di depan/belakang dan spasi ganda
for kol in ["judul", "artis", "genre"]:
    df[kol] = df[kol].str.strip().str.replace(r"\s+", " ", regex=True)


# judul/artis yang semuanya huruf besar atau semuanya kecil dirapikan jadi huruf kapital di awal kata
# (nama pendek seperti BTS, PSY, ABBA sengaja dibiarkan karena memang singkatan)
def rapikan_judul(j):
    if j.isupper() or j.islower():
        return j.title()
    return j


def rapikan_artis(a):
    if (a.isupper() and len(a) > 4) or a.islower():
        return a.title()
    return a


df["judul"] = df["judul"].apply(rapikan_judul)
df["artis"] = df["artis"].apply(rapikan_artis)

# genre: samakan penulisan lewat kamus
peta_genre = {
    "pop": "Pop",
    "rock": "Rock",
    "hip hop": "Hip-Hop",
    "hip-hop": "Hip-Hop",
    "kpop": "K-Pop",
    "k-pop": "K-Pop",
    "edm": "EDM",
    "funk": "Funk",
    "latin": "Latin",
    "disco": "Disco",
}
df["genre"] = df["genre"].str.lower().map(peta_genre)

print("Genre setelah distandarkan:")
print(df["genre"].value_counts(dropna=False))

Genre setelah distandarkan:
genre
Pop        16
Rock        7
K-Pop       3
Latin       2
EDM         2
Disco       2
Hip-Hop     1
Funk        1
NaN         1
Name: count, dtype: int64


### 3.2 Perbaikan tipe data dan format

In [15]:
# jumlah_putar: hapus koma ribuan lalu jadikan angka
df["jumlah_putar"] = pd.to_numeric(df["jumlah_putar"].str.replace(",", ""), errors="coerce")

# durasi: samakan semuanya jadi detik
df["durasi_detik"] = df["durasi"].apply(durasi_ke_detik)
df = df.drop(columns="durasi")


# tanggal_rilis: coba tiga format yang ada, kalau gagal dijadikan NaT (kosong)
def parse_tanggal(teks):
    if pd.isna(teks):
        return pd.NaT
    for fmt in ["%Y-%m-%d", "%d/%m/%Y", "%d-%m-%Y"]:
        try:
            return pd.Timestamp(datetime.strptime(teks.strip(), fmt))
        except ValueError:
            continue
    return pd.NaT


df["tanggal_rilis"] = pd.to_datetime(df["tanggal_rilis"].apply(parse_tanggal))

df[["judul", "tanggal_rilis", "durasi_detik", "jumlah_putar"]].head(10)

,judul,tanggal_rilis,durasi_detik,jumlah_putar
0,Shape of You,2017-01-06,233.0,3.100000e+09
1,Blinding Lights,2019-11-29,200.0,3.400000e+09
2,Bohemian Rhapsody,1975-10-31,355.0,2.000000e+09
3,Someone Like You,2011-01-24,285.0,1.700000e+09
4,Lose Yourself,2002-10-28,326.0,2.100000e+09
5,Hotel California,1976-12-08,390.0,1.500000e+09
6,Smells Like Teen Spirit,1991-09-10,301.0,1.800000e+09
7,Billie Jean,1983-01-02,294.0,1.600000e+09
8,Uptown Funk,2014-11-10,270.0,2.300000e+09
9,Despacito,2017-01-12,228.0,3.000000e+09


### 3.3 Menangani data yang tidak valid

Nilai yang tidak mungkin saya ubah jadi kosong (NaN), lalu diisi di langkah missing value nanti.

In [16]:
valid_awal = df[["popularitas", "tahun_rilis", "jumlah_putar", "durasi_detik"]].notna().sum()

df.loc[(df["popularitas"] < 0) | (df["popularitas"] > 100), "popularitas"] = np.nan
df.loc[(df["tahun_rilis"] < 1950) | (df["tahun_rilis"] > 2026), "tahun_rilis"] = np.nan
df.loc[df["jumlah_putar"] <= 0, "jumlah_putar"] = np.nan
df.loc[df["durasi_detik"] < 30, "durasi_detik"] = np.nan

valid_akhir = df[["popularitas", "tahun_rilis", "jumlah_putar", "durasi_detik"]].notna().sum()
print("Nilai tidak valid yang dijadikan kosong:")
print(valid_awal - valid_akhir)

Nilai tidak valid yang dijadikan kosong:
popularitas     2
tahun_rilis     1
jumlah_putar    1
durasi_detik    1
dtype: int64


### 3.4 Menghapus data duplikat

In [17]:
sebelum = len(df)
df = df.drop_duplicates(subset=["judul", "artis"], keep="first").reset_index(drop=True)
print(f"Baris sebelum: {sebelum}, sesudah: {len(df)}, dibuang: {sebelum - len(df)}")

Baris sebelum: 35, sesudah: 33, dibuang: 2


### 3.5 Menangani outlier

Sesuai hasil pengecekan di bagian 2.6, hanya nilai yang memang tidak masuk akal yang saya perbaiki. Durasi lebih dari 20 menit dan jumlah putar lebih dari 10 miliar saya anggap salah input, lalu dijadikan kosong supaya nanti diisi median.

In [18]:
print("Outlier durasi:")
print(df.loc[df["durasi_detik"] > 1200, ["judul", "durasi_detik"]])
print()
print("Outlier jumlah putar:")
print(df.loc[df["jumlah_putar"] > 10_000_000_000, ["judul", "jumlah_putar"]])

df.loc[df["durasi_detik"] > 1200, "durasi_detik"] = np.nan
df.loc[df["jumlah_putar"] > 10_000_000_000, "jumlah_putar"] = np.nan

Outlier durasi:
     judul  durasi_detik
15  Kangen        3500.0

Outlier jumlah putar:
            judul  jumlah_putar
29  Gangnam Style  9.500000e+10


### 3.6 Menangani missing value

Cara yang saya pilih untuk tiap kolom:
- `tahun_rilis`: diisi dari tahun pada `tanggal_rilis` (karena dua kolom itu berhubungan). Kalau tanggal juga kosong, baru dipakai median.
- `genre`: diisi `Tidak Diketahui`, karena genre tidak bisa ditebak dengan median/rata-rata.
- `popularitas`, `jumlah_putar`, `durasi_detik`: diisi median. Saya pilih median (bukan rata-rata) karena nilainya bisa berat sebelah dan median tidak terpengaruh nilai ekstrem.
- `tanggal_rilis`: dibiarkan kosong. Tahunnya sudah ada di kolom `tahun_rilis` dan tanggal pastinya tidak dipakai di analisis, jadi tidak perlu ditebak-tebak.

In [19]:
print("Missing value sebelum diisi:")
print(df.isna().sum()[df.isna().sum() > 0])

df["tahun_rilis"] = df["tahun_rilis"].fillna(df["tanggal_rilis"].dt.year)
df["tahun_rilis"] = df["tahun_rilis"].fillna(df["tahun_rilis"].median())
df["genre"] = df["genre"].fillna("Tidak Diketahui")

for kol in ["popularitas", "jumlah_putar", "durasi_detik"]:
    df[kol] = df[kol].fillna(df[kol].median())

# setelah tidak ada yang kosong, tipe datanya dirapikan jadi bilangan bulat
df["tahun_rilis"] = df["tahun_rilis"].astype(int)
df["popularitas"] = df["popularitas"].round().astype(int)
df["jumlah_putar"] = df["jumlah_putar"].round().astype("int64")
df["durasi_detik"] = df["durasi_detik"].round().astype(int)

print()
print("Missing value sesudah diisi:")
print(df.isna().sum()[df.isna().sum() > 0])

Missing value sebelum diisi:
genre            1
tahun_rilis      2
tanggal_rilis    2
jumlah_putar     3
popularitas      4
durasi_detik     2
dtype: int64

Missing value sesudah diisi:
tanggal_rilis    2
dtype: int64


### 3.7 Pengecekan hasil cleansing

In [20]:
print("Jumlah baris:", len(df))
print("Duplikat:", df.duplicated(subset=["judul", "artis"]).sum())
print()
df.info()

Jumlah baris: 33
Duplikat: 0

<class 'pandas.DataFrame'>
RangeIndex: 33 entries, 0 to 32
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id_lagu        33 non-null     int64         
 1   judul          33 non-null     str           
 2   artis          33 non-null     str           
 3   genre          33 non-null     str           
 4   tahun_rilis    33 non-null     int64         
 5   tanggal_rilis  31 non-null     datetime64[us]
 6   jumlah_putar   33 non-null     int64         
 7   popularitas    33 non-null     int64         
 8   durasi_detik   33 non-null     int64         
dtypes: datetime64[us](1), int64(5), str(3)
memory usage: 2.4 KB


## 4. Data Enrichment

Enrichment artinya menambah informasi baru ke dataset. Ada dua cara yang saya pakai:
1. **Menambah data dari tabel referensi luar.** Saya buat tabel kecil berisi negara asal dan benua tiap artis, lalu digabung (merge) ke dataset berdasarkan nama artis.
2. **Membuat kolom turunan** dari kolom yang sudah ada (dekade, usia lagu, durasi dalam menit, kategori durasi, kategori popularitas, asal lagu, dan rata-rata putar per tahun).

### 4.1 Menambah negara asal dan benua artis (tabel referensi)

In [21]:
tabel_artis = pd.DataFrame(
    [
        ("Ed Sheeran", "Britania Raya", "Eropa"),
        ("The Weeknd", "Kanada", "Amerika Utara"),
        ("Queen", "Britania Raya", "Eropa"),
        ("Adele", "Britania Raya", "Eropa"),
        ("Eminem", "Amerika Serikat", "Amerika Utara"),
        ("Eagles", "Amerika Serikat", "Amerika Utara"),
        ("Nirvana", "Amerika Serikat", "Amerika Utara"),
        ("Michael Jackson", "Amerika Serikat", "Amerika Utara"),
        ("Mark Ronson", "Britania Raya", "Eropa"),
        ("Luis Fonsi", "Puerto Riko", "Amerika Utara"),
        ("Imagine Dragons", "Amerika Serikat", "Amerika Utara"),
        ("Billie Eilish", "Amerika Serikat", "Amerika Utara"),
        ("BTS", "Korea Selatan", "Asia"),
        ("Weird Genius", "Indonesia", "Asia"),
        ("Tulus", "Indonesia", "Asia"),
        ("Dewa 19", "Indonesia", "Asia"),
        ("Coldplay", "Britania Raya", "Eropa"),
        ("Taylor Swift", "Amerika Serikat", "Amerika Utara"),
        ("Bruno Mars", "Amerika Serikat", "Amerika Utara"),
        ("Dua Lipa", "Britania Raya", "Eropa"),
        ("Harry Styles", "Britania Raya", "Eropa"),
        ("Rihanna", "Barbados", "Amerika Utara"),
        ("Maroon 5", "Amerika Serikat", "Amerika Utara"),
        ("Linkin Park", "Amerika Serikat", "Amerika Utara"),
        ("Radiohead", "Britania Raya", "Eropa"),
        ("Sheila on 7", "Indonesia", "Asia"),
        ("ABBA", "Swedia", "Eropa"),
        ("Avicii", "Swedia", "Eropa"),
        ("Daft Punk", "Prancis", "Eropa"),
        ("PSY", "Korea Selatan", "Asia"),
        ("Blackpink", "Korea Selatan", "Asia"),
        ("Lewis Capaldi", "Britania Raya", "Eropa"),
        ("Noah", "Indonesia", "Asia"),
    ],
    columns=["artis", "negara_artis", "benua"],
)

df = df.merge(tabel_artis, on="artis", how="left")
print("Artis yang tidak ketemu di tabel referensi:", df["negara_artis"].isna().sum())
df[["judul", "artis", "negara_artis", "benua"]].head(8)

Artis yang tidak ketemu di tabel referensi: 0


,judul,artis,negara_artis,benua
0,Shape of You,Ed Sheeran,Britania Raya,Eropa
1,Blinding Lights,The Weeknd,Kanada,Amerika Utara
2,Bohemian Rhapsody,Queen,Britania Raya,Eropa
3,Someone Like You,Adele,Britania Raya,Eropa
4,Lose Yourself,Eminem,Amerika Serikat,Amerika Utara
5,Hotel California,Eagles,Amerika Serikat,Amerika Utara
6,Smells Like Teen Spirit,Nirvana,Amerika Serikat,Amerika Utara
7,Billie Jean,Michael Jackson,Amerika Serikat,Amerika Utara


### 4.2 Membuat kolom turunan

In [22]:
TAHUN_SEKARANG = 2026

# dekade rilis, misal 2017 -> 2010-an
df["dekade"] = (df["tahun_rilis"] // 10 * 10).astype(str) + "-an"

# usia lagu dalam tahun
df["usia_lagu"] = TAHUN_SEKARANG - df["tahun_rilis"]

# durasi dalam menit
df["durasi_menit"] = (df["durasi_detik"] / 60).round(2)

# kategori durasi: pendek (sampai 3:30), sedang (sampai 5:00), panjang (lebih dari 5:00)
df["kategori_durasi"] = pd.cut(
    df["durasi_detik"], bins=[0, 210, 300, np.inf], labels=["Pendek", "Sedang", "Panjang"]
).astype(str)

# kategori popularitas: rendah (<70), sedang (70-84), tinggi (85 ke atas)
df["kategori_popularitas"] = pd.cut(
    df["popularitas"], bins=[-1, 69, 84, 100], labels=["Rendah", "Sedang", "Tinggi"]
).astype(str)

# lagu lokal (artis Indonesia) atau internasional
df["asal_lagu"] = np.where(df["negara_artis"] == "Indonesia", "Lokal", "Internasional")

# rata-rata jumlah putar per tahun sejak rilis
df["putar_per_tahun"] = (df["jumlah_putar"] / df["usia_lagu"].clip(lower=1)).round().astype("int64")

df[["judul", "dekade", "usia_lagu", "durasi_menit", "kategori_durasi", "kategori_popularitas", "asal_lagu", "putar_per_tahun"]].head(10)

,judul,dekade,usia_lagu,durasi_menit,kategori_durasi,kategori_popularitas,asal_lagu,putar_per_tahun
0,Shape of You,2010-an,9,3.88,Sedang,Tinggi,Internasional,344444444
1,Blinding Lights,2010-an,7,3.33,Pendek,Tinggi,Internasional,485714286
2,Bohemian Rhapsody,1970-an,51,5.92,Panjang,Tinggi,Internasional,39215686
3,Someone Like You,2010-an,15,4.75,Sedang,Sedang,Internasional,113333333
4,Lose Yourself,2000-an,24,5.43,Panjang,Tinggi,Internasional,87500000
5,Hotel California,1970-an,50,6.50,Panjang,Sedang,Internasional,30000000
6,Smells Like Teen Spirit,1990-an,35,5.02,Panjang,Sedang,Internasional,51428571
7,Billie Jean,1980-an,43,4.90,Sedang,Sedang,Internasional,37209302
8,Uptown Funk,2010-an,12,4.50,Sedang,Sedang,Internasional,191666667
9,Despacito,2010-an,9,3.80,Sedang,Sedang,Internasional,333333333


In [23]:
print("Kolom sekarang:", len(df.columns))
print(list(df.columns))
df.head()

Kolom sekarang: 18
['id_lagu', 'judul', 'artis', 'genre', 'tahun_rilis', 'tanggal_rilis', 'jumlah_putar', 'popularitas', 'durasi_detik', 'negara_artis', 'benua', 'dekade', 'usia_lagu', 'durasi_menit', 'kategori_durasi', 'kategori_popularitas', 'asal_lagu', 'putar_per_tahun']


,id_lagu,judul,artis,genre,tahun_rilis,tanggal_rilis,jumlah_putar,popularitas,durasi_detik,negara_artis,benua,dekade,usia_lagu,durasi_menit,kategori_durasi,kategori_popularitas,asal_lagu,putar_per_tahun
0,1,Shape of You,Ed Sheeran,Pop,2017,2017-01-06,3100000000,90,233,Britania Raya,Eropa,2010-an,9,3.88,Sedang,Tinggi,Internasional,344444444
1,2,Blinding Lights,The Weeknd,Pop,2019,2019-11-29,3400000000,92,200,Kanada,Amerika Utara,2010-an,7,3.33,Pendek,Tinggi,Internasional,485714286
2,3,Bohemian Rhapsody,Queen,Rock,1975,1975-10-31,2000000000,88,355,Britania Raya,Eropa,1970-an,51,5.92,Panjang,Tinggi,Internasional,39215686
3,4,Someone Like You,Adele,Pop,2011,2011-01-24,1700000000,80,285,Britania Raya,Eropa,2010-an,15,4.75,Sedang,Sedang,Internasional,113333333
4,5,Lose Yourself,Eminem,Hip-Hop,2002,2002-10-28,2100000000,85,326,Amerika Serikat,Amerika Utara,2000-an,24,5.43,Panjang,Tinggi,Internasional,87500000


## 5. Perbandingan Kondisi Data Sebelum dan Sesudah

### 5.1 Tabel perbandingan

In [24]:
def hitung_tidak_valid(d_pop, d_tahun, d_putar, d_detik):
    return int(((d_pop < 0) | (d_pop > 100)).sum()
               + ((d_tahun < 1950) | (d_tahun > 2026)).sum()
               + (d_putar <= 0).sum()
               + (d_detik < 30).sum())


tidak_valid_sebelum = hitung_tidak_valid(
    df_mentah["popularitas"], df_mentah["tahun_rilis"], putar_cek, detik_cek
) + 1  # +1 untuk tanggal 31/02/1999

outlier_sebelum = int((detik_cek > 1200).sum() + (putar_cek > 10_000_000_000).sum())

perbandingan = pd.DataFrame(
    {
        "Sebelum": {
            "Jumlah baris": len(df_mentah),
            "Jumlah kolom": df_mentah.shape[1],
            "Total missing value": int(df_mentah.isna().sum().sum()),
            "Baris duplikat": int(kunci.duplicated().sum()),
            "Variasi penulisan genre": df_mentah["genre"].nunique(),
            "Nilai tidak valid": tidak_valid_sebelum,
            "Outlier": outlier_sebelum,
            "Kolom dengan tipe data keliru": 3,  # jumlah_putar, durasi, tanggal_rilis (terbaca teks)
        },
        "Sesudah": {
            "Jumlah baris": len(df),
            "Jumlah kolom": df.shape[1],
            "Total missing value": int(df.isna().sum().sum()),
            "Baris duplikat": int(df.duplicated(subset=["judul", "artis"]).sum()),
            "Variasi penulisan genre": df["genre"].nunique(),
            "Nilai tidak valid": hitung_tidak_valid(
                df["popularitas"], df["tahun_rilis"], df["jumlah_putar"], df["durasi_detik"]
            ),
            "Outlier": int((df["durasi_detik"] > 1200).sum() + (df["jumlah_putar"] > 10_000_000_000).sum()),
            "Kolom dengan tipe data keliru": 0,  # sudah angka dan tanggal semua
        },
    }
)
perbandingan

,Sebelum,Sesudah
Jumlah baris,35,33
Jumlah kolom,9,18
Total missing value,6,2
Baris duplikat,2,0
Variasi penulisan genre,14,9
Nilai tidak valid,6,0
Outlier,2,0
Kolom dengan tipe data keliru,3,0


Catatan:
- Pada kondisi sesudah masih ada 2 sel kosong di `tanggal_rilis` (Sunday Morning dan Sephia). Itu memang sengaja tidak diisi, alasannya sudah dijelaskan di bagian 3.6.
- Jumlah baris turun dari 35 jadi 33 karena 2 baris duplikat dibuang.
- Variasi genre sesudah cleansing ada 9: 8 genre asli ditambah satu label `Tidak Diketahui` untuk lagu yang genrenya kosong.
- Jumlah kolom naik dari 9 jadi 18 karena ada 9 kolom hasil enrichment (kolom `durasi` diganti `durasi_detik`, jadi jumlahnya tetap 9 sebelum enrichment).

### 5.2 Contoh data sebelum dan sesudah

In [25]:
contoh = ["Blinding Lights", "Despacito", "Kangen", "Levitating", "Umbrella", "Gangnam Style"]

print("SEBELUM (data mentah):")
display(df_mentah[df_mentah["judul"].str.strip().str.title().isin(contoh)][["judul", "artis", "genre", "tahun_rilis", "durasi", "jumlah_putar", "popularitas"]])

print("SESUDAH (data bersih):")
display(df[df["judul"].isin(contoh)][["judul", "artis", "genre", "tahun_rilis", "durasi_detik", "jumlah_putar", "popularitas"]])

SEBELUM (data mentah):


,judul,artis,genre,tahun_rilis,durasi,jumlah_putar,popularitas
1,Blinding Lights,THE WEEKND,pop,2019.0,3:20,"3,400,000,000",92.0
9,Despacito,Luis Fonsi,Latin,2017.0,228,"3,000,000,000",78.0
15,Kangen,Dewa 19,Rock,1992.0,58:20,"410,000,000",70.0
19,Levitating,Dua Lipa,Pop,2035.0,3:23,"2,500,000,000",88.0
21,Umbrella,Rihanna,Pop,2007.0,4:36,"1,500,000,000",150.0
29,Gangnam Style,PSY,K-Pop,2012.0,3:39,"95,000,000,000",77.0
34,DESPACITO,luis fonsi,latin,2017.0,228,"3,000,000,000",78.0


SESUDAH (data bersih):


,judul,artis,genre,tahun_rilis,durasi_detik,jumlah_putar,popularitas
1,Blinding Lights,The Weeknd,Pop,2019,200,3400000000,92
9,Despacito,Luis Fonsi,Latin,2017,228,3000000000,78
15,Kangen,Dewa 19,Rock,1992,233,410000000,70
19,Levitating,Dua Lipa,Pop,2020,203,2500000000,88
21,Umbrella,Rihanna,Pop,2007,276,1500000000,82
29,Gangnam Style,PSY,K-Pop,2012,219,1800000000,77


### 5.3 Insight singkat

In [26]:
print("Rata-rata popularitas per genre:")
print(df.groupby("genre")["popularitas"].mean().round(1).sort_values(ascending=False))
print()
print("Lagu lokal vs internasional:")
print(df.groupby("asal_lagu")[["popularitas", "jumlah_putar", "durasi_menit"]].mean().round(1))
print()
print("Jumlah lagu per dekade:")
print(df["dekade"].value_counts().sort_index())

Rata-rata popularitas per genre:
genre
Hip-Hop            85.0
Tidak Diketahui    84.0
Pop                83.1
Funk               81.0
Rock               80.9
K-Pop              80.0
Disco              79.5
EDM                78.5
Latin              78.0
Name: popularitas, dtype: float64

Lagu lokal vs internasional:
               popularitas  jumlah_putar  durasi_menit
asal_lagu                                             
Internasional         83.1  2.003571e+09           4.1
Lokal                 73.8  4.520000e+08           3.9

Jumlah lagu per dekade:
dekade
1970-an     3
1980-an     1
1990-an     4
2000-an     5
2010-an    15
2020-an     5
Name: count, dtype: int64


Dari data yang sudah bersih dan diperkaya:

- Genre **Pop** paling banyak (16 dari 33 lagu) dengan rata-rata popularitas 83,1. Hip-Hop memang tertinggi (85,0), tapi isinya cuma 1 lagu jadi kurang bisa dijadikan patokan.
- Lagu **internasional** rata-rata lebih populer (83,1) dibanding lagu **lokal** (73,8), dan jumlah putarnya jauh lebih besar (sekitar 2,0 miliar dibanding 452 juta). Durasinya hampir sama, sekitar 4 menit.
- Paling banyak lagu berasal dari dekade **2010-an** (15 lagu).

Datanya hanya 33 lagu dan angkanya contoh, jadi ini sebatas gambaran untuk latihan.

## 6. Kesimpulan

Dataset lagu yang saya pakai awalnya punya 35 baris dengan beberapa masalah kualitas data: 6 sel kosong, 2 baris duplikat, penulisan genre yang tidak seragam (14 penulisan untuk 8 genre), format durasi dan tanggal yang campur-campur, 3 kolom dengan tipe data keliru, 6 nilai yang tidak valid, dan 2 outlier.

Teknik cleansing yang saya terapkan:
- standardisasi teks (spasi, huruf besar/kecil, dan kamus genre),
- konversi tipe data (jumlah putar jadi angka, durasi jadi detik, tanggal jadi datetime),
- nilai tidak valid dijadikan kosong lalu diisi,
- penghapusan duplikat,
- penanganan outlier dengan metode IQR yang dicek manual,
- pengisian missing value (dari kolom terkait, label `Tidak Diketahui`, dan median).

Untuk enrichment, saya menambahkan negara asal dan benua artis dari tabel referensi, serta kolom turunan berupa dekade, usia lagu, durasi menit, kategori durasi, kategori popularitas, asal lagu, dan rata-rata putar per tahun.

Hasilnya, data jadi 33 baris yang konsisten, tanpa duplikat, tanpa nilai tidak valid, dan hanya menyisakan 2 sel kosong di `tanggal_rilis` yang memang sengaja tidak diisi. Kolomnya bertambah dari 9 jadi 18 sehingga datanya lebih kaya untuk dianalisis.

Yang saya pelajari dari tugas ini: urutan pembersihan itu penting (misalnya duplikat dibuang sebelum mengisi median), dan nilai yang tertandai outlier tidak otomatis salah, jadi tetap perlu dicek dengan logika (misalnya lagu 6 menit masih wajar, tapi lagu 58 menit tidak).

## 7. Menyimpan Hasil

In [27]:
df_mentah.to_csv("dataset_musik_mentah.csv", index=False)
df.to_csv("dataset_musik_bersih.csv", index=False)
print("File tersimpan: dataset_musik_mentah.csv dan dataset_musik_bersih.csv")

File tersimpan: dataset_musik_mentah.csv dan dataset_musik_bersih.csv
